task-maker-async/src/store.rs

```diff
@@ -0,0 +1,114 @@
+#![allow(dead_code)]
+
+use serde::{Deserialize, Serialize};
+
+type HashData = Vec<u8>;
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct InputFileHash(HashData);
+
+/// Two-level hash; the outer hash has information about all properties of a computation that are
+/// expected to change the result (such as outer hashes of inputs, or the command line). The inner
+/// hash takes care of properties of the computation that should not change the outputs, such as
+/// time and memory limits.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ComputationHash(HashData, HashData);
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct InputFileWriteHandle(usize);
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ComputationWriteHandle(usize);
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ReadHandle(usize);
+
+#[derive(Debug, Serialize, Deserialize)]
+pub enum ComputationDataFile {
+    Outcome,
+    OutputFile(String),
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub enum FileReadingOutcome {
+    /// The file has been deleted, for example because the worker responsible for the execution has
+    /// disappeared.
+    Dropped,
+    /// The file has been fully read.
+    EndOfFile,
+    /// A new chunk of data is available.
+    Data(Vec<u8>),
+}
+
+#[tarpc::service]
+pub trait Store {
+    /// Used by the client to upload inputs to the DAG. Returns None if the file is already
+    /// present. Either way, this leases the corresponding hash for LEASE_LENGTH_SEC.
```

Comment:
Present and finalized?

Reply:
Right. What do you suggest we should do if present and *not* finalized? Return an error?

Reply:
(this is how I modified the API for now)

Reply:
Yes, I think that's fine. I'm thinking of leases being quite short (O(second)), so this shouldn't introduce noticeable delays.